engine/utils/String.h

```diff
@@ -610,5 +626,19 @@ String operator+(const char& lhs, const String& rhs);
  * @return the formatted std::ostream instance
  */
 std::ostream& operator<<(std::ostream& os, const String& string);
+} // namespace Siege
+
+// Define a hash function for the string so that it can be stored in a map or ordered set
+// TODO(Aryeh): Maybe replace this with a custom implementation when the time comes.
+template<>
+struct std::hash<Siege::String>
+{
+    size_t operator()(const Siege::String& str) const
+    {
+        size_t seed = 0;
+        Siege::Hash::HashCombine(seed, str.At(0), str.At(str.Size()), str.Size(), str.Capacity());
+        return seed;
+    };
+};
```

Comment:
Great! However I'm wondering why this was needed suddenly? Were we not using strings as keys in maps before this PR? Also, please change parameter name to `string` in accordance with the rest of the interface.

Reply:
This isn't suddenly needed. The material system uses this to compare uniform names. This is an optimisation which ensures that we don't need to compare strings while doing so every frame.